File: Hoare Partition.py
# code from https://www.geeksforgeeks.org/hoares-vs-lomuto-partition-scheme-quicksort/


def partition(arr, low, high):
    pivot = arr[low]
    i = low - 1
    j = high + 1

    while (True):

		# Find leftmost element greater than
		# or equal to pivot
        i += 1
        while (arr[i] < pivot):
            i += 1

		# Find rightmost element smaller than
		# or equal to pivot
        j -= 1
        while (arr[j] > pivot):
            j -= 1

        # If two pointers met.
        if (i >= j):
            return j

        arr[i], arr[j] = arr[j], arr[i]


''' The main function that implements QuickSort
arr --> Array to be sorted,
low --> Starting index,
high --> Ending index '''


def quickSort(arr, low, high):
	''' pi is partitioning index, arr[p] is now
	at right place '''
	if (low < high):

		pi = partition(arr, low, high)

		# Separately sort elements before
		# partition and after partition
		quickSort(arr, low, pi)
		quickSort(arr, pi + 1, high)


from openpyxl import Workbook

wb = Workbook()
ws = wb.active
ws['A1'] = '時間'

import numpy as np
import time



for a in range(10, 26):
    n = 2 ** a
    ws.append(["2的", a])
    for k in range(10):

        random_list = np.random.randint(0, 1001, size=n)
        # print("Given array is")
        # for x in range(n):
        #     print("%d" % random_list[x]),
        start = time.time()
        quickSort(random_list, 0, n-1)
        end = time.time()
        # print("\n\nSorted array is")
        # for i in range(n):
        #     print("%d" % random_list[i]),
        print("use ", end-start, "time")

        ws.append([end-start])
wb.save("hoare.xlsx")

File: Hoare Partition.cc
//https://www.geeksforgeeks.org/hoares-vs-lomuto-partition-scheme-quicksort/?ref=gcse
/* C++ implementation of QuickSort using Hoare's
   partition scheme. */
//#include <bits/stdc++.h>
#include <iostream>
#include <vector>
#include <cmath>
#include <cstdlib>
#include <ctime>
using namespace std;

/* This function takes first element as pivot, and places
   all the elements smaller than the pivot on the left side
   and all the elements greater than the pivot on
   the right side. It returns the index of the last element
   on the smaller side*/
int partition(vector<int> arr, int low, int high)
{
    int pivot = arr[low];
    int i = low - 1, j = high + 1;

    while (true) {
        // Find leftmost element greater than
        // or equal to pivot
        do {
            i++;
        } while (arr[i] < pivot);

        // Find rightmost element smaller than
        // or equal to pivot
        do {
            j--;
        } while (arr[j] > pivot);

        // If two pointers met.
        if (i >= j)
            return j;

        swap(arr[i], arr[j]);
    }
}

/* The main function that implements QuickSort
   arr[] --> Array to be sorted,
   low --> Starting index,
   high --> Ending index */
void quickSort(vector<int> arr, int low, int high)
{
    if (low < high) {
        /* pi is partitioning index, arr[p] is now
           at right place */
        int pi = partition(arr, low, high);

        // Separately sort elements before
        // partition and after partition
        quickSort(arr, low, pi);
        quickSort(arr, pi + 1, high);
    }
}

/* Function to print an array */
void printArray(int arr[], int n)
{
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

// Driver Code
int main()
{
       for(int k = 15; k<=30;k++){
          float T = 0;
          for(int t = 0; t < 10; t++){
              int SIZEOFARR = pow(2,k);
              vector <int> Myvector;
              for (int i = 0; i< SIZEOFARR; i++){
                  Myvector.push_back( rand()%1001 );
              }
              auto arrsize = sizeof(Myvector) / sizeof(Myvector[0]);
              clock_t Startime = clock();
              quickSort(Myvector,0,arrsize-1);
              clock_t Endtime = clock();
              T += (float)(Endtime - Startime)/CLOCKS_PER_SEC;
              Myvector.clear();
          }
          printf("When k = %d, It takes %.2f (s).\n",k ,T/10);
      }
      return 0;
}


